### PR TITLE
New test case : Verify Enable-Disable SRIOV.

### PR DESCRIPTION
**Pull request by iamshital**
Azure now supports Enable/Disable Accelerated networking for existing VMs. This means, we do not need to delete and recreate the VMs to alter SRIOV status.

### CODE REVIEW

**Comment by iamshital** [comment]
Execution on local machine is verified. 
Jenkins execution is in progress...  (job-cloudtest-auto/5572)

File: Libraries/Azure.psm1 [hunk 1]
```diff
@@ -2540,4 +2540,143 @@ Function RestartAllAzureDeployments($allVMData)
 	Remove-Item -Path "$env:TEMP\$($currentGUID).azurecontext" -Force -ErrorAction SilentlyContinue | Out-Null
 	$isSSHOpened = isAllSSHPortsEnabledRG -AllVMDataObject $AllVMData
 	return $isSSHOpened
+}
+
+Function Enable-SRIOVinAzureVM($ResourceGroup, $VMName) {
```
Comment (juhlee-microsoft) [inline Libraries/Azure.psm1]: Can you merge those two functions into one with flag? Also any chance for supporting Hyper-V also in the same function?

File: Libraries/CommonFunctions.psm1 [hunk 2]
```diff
@@ -3938,4 +3939,38 @@ function Get-NumaSupportStatus {
     # We skip the check if kernel is not 2.6
     # Anything newer will have support for it
     return $true
+}
+
+Function Test-SRIOVInLinuxGuest {
```
Comment (juhlee-microsoft) [inline Libraries/CommonFunctions.psm1]: Get-SRIOVState?

File: Testscripts/Windows/VERIFY-ENABLE-DISABLE-SRIOV-ON-EXISTING-VM.ps1 [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+                    -checkValues "PASS,FAIL,ABORTED" -testName $currentTestData.testName                    
+                }
+            }
+            else {
+                #Break the for loop.
+                $resultArr += "FAIL"
+                break;
+            }
+
+            if ($Stage1Result) {
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/VERIFY-ENABLE-DISABLE-SRIOV-ON-EXISTING-VM.ps1]: it does not look it should be out here. 'true' part can be inserted at the line 22.

File: Testscripts/Windows/VERIFY-ENABLE-DISABLE-SRIOV-ON-EXISTING-VM.ps1 [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            }
+        }
+    }
+    catch {
+        $ErrorMessage = $_.Exception.Message
+        LogErr "EXCEPTION : $ErrorMessage"
+    }
+    Finally {
+        if (!$testResult) {
+            $testResult = "Aborted"
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/VERIFY-ENABLE-DISABLE-SRIOV-ON-EXISTING-VM.ps1]: ABORTED

File: Testscripts/Windows/VERIFY-ENABLE-DISABLE-SRIOV-ON-EXISTING-VM.ps1 [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            }
+            else {
+                #Break the for loop.
+                $resultArr += "FAIL"
+                break;
+            }
+        }
+    }
+    catch {
+        $ErrorMessage = $_.Exception.Message
```
Comment (juhlee-microsoft) [inline Testscripts/Windows/VERIFY-ENABLE-DISABLE-SRIOV-ON-EXISTING-VM.ps1]: Add the error line number

File: Libraries/Azure.psm1 (hunk 1 above)
Comment (iamshital) [reply Libraries/Azure.psm1]: I had this thought actually. Like -
```
Manage-SRIOVinAzureVM -ResourceGroup <ResourceGroup> -VMName <VMName> -Enable
Manage-SRIOVinAzureVM -ResourceGroup <ResourceGroup> -VMName <VMName> -Disable
```
But, keeping them separate looked good to me.
For example, how PowerShell is managing it's default functions.
```
Install-Module -Module <ModuleName>
Uninstall-Module -Module <ModuleName>
```
It feels like we're trying to merge them into following -
```
Manage-Module -ModuleName <ModuleName> -Install
Manage-Module -ModuleName <ModuleName> -Install
```
Which didn't look good actually.

File: Testscripts/Windows/VERIFY-ENABLE-DISABLE-SRIOV-ON-EXISTING-VM.ps1 (hunk 3 above)
Comment (iamshital) [reply Testscripts/Windows/VERIFY-ENABLE-DISABLE-SRIOV-ON-EXISTING-VM.ps1]: It is inserted at line no 19 and we're verifying it at line no 45.

**Comment by iamshital** [comment]
Execution is verified.
----------------
```
09/06/2018 21:07:51 : [INFO ] ~~~~~~~~~~~~~~~TEST END : VERIFY-ENABLE-DISABLE-SRIOV-ON-EXISTING-VM~~~~~~~~~~
09/06/2018 21:07:51 : [INFO ] CURRENT - PASS    - 1
09/06/2018 21:07:51 : [INFO ] CURRENT - FAIL    - 0
09/06/2018 21:07:51 : [INFO ] CURRENT - ABORTED - 0
09/06/2018 21:07:51 : [INFO ] Checking background cleanup jobs..
09/06/2018 21:07:51 : [INFO ] Removed 0 context files.
09/06/2018 21:07:51 : [INFO ] Cycle Finished.. TC-44786
09/06/2018 21:07:51 : [INFO ] 
Test Results Summary
ICA test run on 09/06/2018 20:28:55
ARM Image under test Canonical : UbuntuServer : 16.04-LTS : latest
Total Executed TestCases        1 (       1 Pass,        0 Fail,        0 Abort)
Total Execution Time(dd:hh:mm) 0:0:38
XML file: TestConfiguration

VERIFY-ENABLE-DISABLE-SRIOV-ON-EXISTING-VM Execution Time: 38.9 minutes
	VERIFY-ENABLE-DISABLE-SRIOV-ON-EXISTING-VM : PASS  
		  EnableSRIOV : Test Iteration - 1 : PASS 
		  DisableSRIOV : Test Iteration - 1 : PASS 
		  EnableSRIOV : Test Iteration - 2 : PASS 
		  DisableSRIOV : Test Iteration - 2 : PASS 
		  EnableSRIOV : Test Iteration - 3 : PASS 
		  DisableSRIOV : Test Iteration - 3 : PASS 
		  EnableSRIOV : Test Iteration - 4 : PASS 
		  DisableSRIOV : Test Iteration - 4 : PASS 
		  EnableSRIOV : Test Iteration - 5 : PASS 
		  DisableSRIOV : Test Iteration - 5 : PASS 


Logs can be found at .\TestResults\2018-06-09-20-28-49-4095


09/06/2018 21:07:51 : [INFO ] Creating 'Azure-44786-buildlogs.zip' from '.\TestResults\2018-06-09-20-28-49-4095'
09/06/2018 21:07:51 : [INFO ] C:\LISAv2\PLBR44786\Azure-44786-buildlogs.zip created successfully.
09/06/2018 21:07:51 : [INFO ] Copied : .\report\report_TC-44786.xml --> .\report\report_TC-44786-junit.xml
09/06/2018 21:07:51 : [INFO ] Analysing results..
09/06/2018 21:07:51 : [INFO ] PASS  : 1
09/06/2018 21:07:52 : [INFO ] FAIL  : 0
09/06/2018 21:07:52 : [INFO ] ABORT : 0
Copying all files back to original working directory: C:\Jenkins-Slaves\azure-slave-2-executor-7\workspace\job-azuretest\job-cloudtest-auto.
Cleaning up C:\LISAv2\PLBR44786
Setting workspace back to original location: C:\Jenkins-Slaves\azure-slave-2-executor-7\workspace\job-azuretest\job-cloudtest-auto
09/06/2018 21:07:54 : [INFO ] LISAv2 exits with code: 0
```

File: Testscripts/Windows/VERIFY-ENABLE-DISABLE-SRIOV-ON-EXISTING-VM.ps1 (hunk 3 above)
Comment (juhlee-microsoft) [reply Testscripts/Windows/VERIFY-ENABLE-DISABLE-SRIOV-ON-EXISTING-VM.ps1]: The line number 45's 'true' part is only executible when it reach to the line number 19.  I suggest you put this 'true' part below the line number 19.

File: Libraries/Azure.psm1 (hunk 1 above)
Comment (iamshital) [reply Libraries/Azure.psm1]: @juhlee-microsoft , I agree there are some duplicate lines among those two functions.

I have two options -
1. Make single function with -Enable / -Disable toggle. (In this case, please also suggest a Function Name)
2. Keep the functions separate and as it is, but handle the duplicate code in some optimal way.
Let me know your thoughts.

Comment (juhlee-microsoft) [reply Libraries/Azure.psm1]: 1 & Set-SRIOVInVM. Omitted 'Azure', plan to add Hyper-V part in future.

Comment (iamshital) [reply Libraries/Azure.psm1]: Sounds good. Will do this.

**Comment by juhlee-microsoft** [comment]
Please follow up each open comment.

File: Libraries/CommonFunctions.psm1 [hunk 6]
```diff
@@ -2237,8 +2237,9 @@ Function GetAllDeployementData($ResourceGroups)
 			$QuickVMNode.Location = $currentRGLocation
 			$allDeployedVMs += $QuickVMNode
 		}
-		LogMsg "Collected $ResourceGroup data!"		
+		LogMsg "Collected $ResourceGroup data!"
 	}
+	Set-Variable -Name AllVMData -Value $allDeployedVMs -Scope Global
```
[the remark this answers is not in the document]
Comment (iamshital) [reply Libraries/CommonFunctions.psm1]: It is necessary.
Public IP changes every time we shutdown the VM. Hence, we need to refresh the data and make new information available globally.

File: Testscripts/Windows/VERIFY-ENABLE-DISABLE-SRIOV-ON-EXISTING-VM.ps1 [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+# Licensed under the Apache License.
+
+function Main {
+    try {
+        $resultArr = @()
+        $Stage2Result = $true
+        $SuperUser = "root"
+        $FailureCount = 0
+        #Enable SRIOV
+        for ($TestIteration = 1 ; $TestIteration -le [int]$CurrentTestData.TestIterations; $TestIteration ++ ) {
```
[the remark this answers is not in the document]
Comment (iamshital) [reply Testscripts/Windows/VERIFY-ENABLE-DISABLE-SRIOV-ON-EXISTING-VM.ps1]: Fixed.

**Comment by iamshital** [comment]
@juhlee-microsoft , @lizzha  comments are addressed.

**Pull request merged**